swarm/src/handler.rs

```diff
@@ -222,6 +222,42 @@ pub enum ConnectionEvent<'a, IP: InboundUpgradeSend, OP: OutboundUpgradeSend, IO
     RemoteProtocolsChange(ProtocolsChange<'a>),
 }
 
+impl<'a, IP, OP, IOI, OOI> fmt::Debug for ConnectionEvent<'a, IP, OP, IOI, OOI>
+where
+    IP: InboundUpgradeSend + fmt::Debug,
+    IP::Output: fmt::Debug,
+    IP::Error: fmt::Debug,
+    OP: OutboundUpgradeSend + fmt::Debug,
+    OP::Output: fmt::Debug,
+    OP::Error: fmt::Debug,
+    IOI: fmt::Debug,
+    OOI: fmt::Debug,
+{
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            ConnectionEvent::FullyNegotiatedInbound(v) => {
+                f.debug_tuple("FullyNegotiatedInbound").field(v).finish()
+            }
+            ConnectionEvent::FullyNegotiatedOutbound(v) => {
+                f.debug_tuple("FullyNegotiatedOutbound").field(v).finish()
+            }
+            ConnectionEvent::AddressChange(v) => f.debug_tuple("AddressChange").field(v).finish(),
+            ConnectionEvent::DialUpgradeError(v) => {
+                f.debug_tuple("DialUpgradeError").field(v).finish()
+            }
+            ConnectionEvent::ListenUpgradeError(v) => {
+                f.debug_tuple("ListenUpgradeError").field(v).finish()
+            }
+            ConnectionEvent::LocalProtocolsChange(v) => {
+                f.debug_tuple("LocalProtocolsChange").field(v).finish()
+            }
+            ConnectionEvent::RemoteProtocolsChange(v) => {
+                f.debug_tuple("RemoteProtocolsChange").field(v).finish()
+            }
+        }
+    }
+}
```

Comment:
Why do we need this? This looks like the impl that would be derived?

Reply:
Derive doesn't know about assosiated types of `IP` and `OP` because there are never used within the enum.  So the following are not produced by it:

```
    IP::Output: fmt::Debug,
    IP::Error: fmt::Debug,
    OP::Output: fmt::Debug,
    OP::Error: fmt::Debug,
```

Reply:
The only way to derive it is by adding the trait bound on the actual struct, but it will restrict the user and make them implement the Debug trait for their type.

In talking about this:

```rust
#[derive(Debug)]
pub enum ConnectionEvent<'a, IP: InboundUpgradeSend, OP: OutboundUpgradeSend, IOI, OOI>
where:
    IP::Output: fmt::Debug,
    IP::Error: fmt::Debug,
    OP::Output: fmt::Debug,
    OP::Error: fmt::Debug,
{
```

Reply:
Thanks for elaborating!